File: tests/rosetta/transpiler/FS/catamorphism.fs
// Generated 2025-07-27 23:36 +0700

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec add (a: int) (b: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    try
        __ret <- a + b
        raise Return
        __ret
    with
        | Return -> __ret
let rec sub (a: int) (b: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    try
        __ret <- a - b
        raise Return
        __ret
    with
        | Return -> __ret
let rec mul (a: int) (b: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    try
        __ret <- a * b
        raise Return
        __ret
    with
        | Return -> __ret
let rec fold (f: int -> int -> int) (xs: int array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable f = f
    let mutable xs = xs
    try
        let mutable r: int = xs.[0]
        let mutable i: int = 1
        while i < (int (Array.length xs)) do
            r <- int (f r (xs.[i]))
            i <- i + 1
        __ret <- r
        raise Return
        __ret
    with
        | Return -> __ret
let n: int array = [|1; 2; 3; 4; 5|]
printfn "%d" (fold (unbox<int -> int -> int> (fun (a: int) (b: int) -> (add a b))) n)
printfn "%d" (fold (unbox<int -> int -> int> (fun (a: int) (b: int) -> (sub a b))) n)
printfn "%d" (fold (unbox<int -> int -> int> (fun (a: int) (b: int) -> (mul a b))) n)
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/rosetta/transpiler/Java/catamorphism.java
public class Main {
    static int[] n;

    static int add(int a, int b) {
        return a + b;
    }

    static int sub(int a, int b) {
        return a - b;
    }

    static int mul(int a, int b) {
        return a * b;
    }

    static int fold(java.util.function.BiFunction<Integer,Integer,Integer> f, int[] xs) {
        int r = xs[0];
        int i = 1;
        while (i < xs.length) {
            r = f.apply(r, xs[i]);
            i = i + 1;
        }
        return r;
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            n = new int[]{1, 2, 3, 4, 5};
            System.out.println(fold((a, b) -> add(a, b), n));
            System.out.println(fold((a_1, b_1) -> sub(a_1, b_1), n));
            System.out.println(fold((a_2, b_2) -> mul(a_2, b_2), n));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }
}
